### PR TITLE
Add functions for listing and downloading archives

### PR DESCRIPTION
**Pull request by as2388**

### CODE REVIEW

File: rapid_pro_tools/rapid_pro_client.py [hunk 1]
```diff
@@ -26,6 +29,68 @@ def __init__(self, server, token):
...
+            for line in decompressed_file.readlines():
+                serialized_run = json.loads(line)
+
+                # Set the 'start' field to null if it doesn't exist.
+                # This field is required to be present in order to be able to deserialize runs, but is often not
+                # present in the downloaded data (possibly because the archiving process removes null fields, but
+                # I haven't verified this). Since this field is often null in the data that comes out of the runs
+                # API directly, and we don't use this in the pipelines, just set missing entries to None.
+                if "start" not in serialized_run:
+                    serialized_run["start"] = None
```
Comment (lukechurch) [inline rapid_pro_tools/rapid_pro_client.py]: This seems odd as a thing to need to do - is it work logging a warning?

File: rapid_pro_tools/rapid_pro_client.py [hunk 2]
```diff
@@ -26,6 +29,68 @@ def __init__(self, server, token):
...
+        :param archive_metadata: Metadata for the archive. To obtain these, see `RapidProClient.list_archives`.
+        :type archive_metadata: temba_client.v2.types.Archive
+        :return: Data downloaded from the archive.
+        :rtype: list of temba_client.v2.Message | list of temba_client.v2.Run
+        """
+        # Download the archive, which is in a gzipped JSONL format, and decompress.
+        log.info(f"Downloading archive from {archive_metadata.download_url}")
+        archive_response = urllib.request.urlopen(archive_metadata.download_url)
+        raw_file = BytesIO(archive_response.read())
+        decompressed_file = gzip.GzipFile(fileobj=raw_file)
```
Comment (lukechurch) [inline rapid_pro_tools/rapid_pro_client.py]: Doing this in memory could use a lot of ram - presumably you've tested this inside a reasonably configured Docker container?

Comment (as2388) [reply rapid_pro_tools/rapid_pro_client.py]: While the memory usage is likely to be high by normal standards, it should be pretty low compared to what the rest of the pipelines need. The largest whole-month archive for IMAQAL is 60MB compressed, so at most a few GB decompressed. I'm not having any problems so far, but will keep this in mind and come back to this if we need to change.

File: rapid_pro_tools/rapid_pro_client.py (hunk 1 above)
Comment (as2388) [reply rapid_pro_tools/rapid_pro_client.py]: I'm not sure what that tells us - I don't know what this field is for, and it's almost always null in the objects read from the runs API too

File: rapid_pro_tools/rapid_pro_client.py (hunk 2 above)
Comment (as2388) [reply rapid_pro_tools/rapid_pro_client.py]: Peak memory use approx. 1GB when loading all the records for CSAP

**Pull request merged**